### ISSUE TITLE
Remove global package installation and make CLI bin command available in the repo

### ISSUE THREAD

**Issue by augnustin** [npm/npm#19355]
#### What's the feature?

- remove `-g` flag from `npm install` command
- make `bin: "path/to/bin"` standard in `package.json` and make CLI command available `npm <PACKAGE NAME>` when `bin` is present (eventually `npm bin <PACKAGE NAME>` but `npm run <COMMAND>` is a pain so shortening feels better)

#### What problem is the feature intended to solve?

As far as I'm concerned, I **never** install packages globally anymore. Why? 

- brings external dependencies 
- makes dev environment uselessly longer to setup (install global dependencies which need to be known, _then_ `npm install`)
- sometimes raises compatibility issues between projects
- in the age of statelessness, this is anti-pattern
- easy to swipe a project: simply remove the directory

Individually, I handle this quite well, define some scripts in `package.json`: `<action>: "node_modules/<PACKAGE_NAME>/bin/<PACKAGE_NAME>.js --options"`. Not very elegant but works.

But when I see the docs suggesting the install packages globally in their getting started (eg. [Browserify](http://browserify.org/), [Parcel](https://parceljs.org/getting_started.html), [Sails.js](https://sailsjs.com/get-started) etc.) I only think of it as an abusing customer retention strategy.

On the other hand, I must say packages that bring CLI tools are convenient and shouldn't be removed of course. Only should they be made accessible in the local context of the repository. Hence the proposal of running `npm <PACKAGE_NAME>` when the `bin: "/path/to/bin"` is defined in the `package.json`.

#### Is the absence of this feature blocking you or your team? If so, how?

Not really. I think this is a bad practice to let people install packages globally.

#### Is this feature similar to an existing feature in another tool?

Well, yarn introduced several differences with npm, but they have left global packages, so the answer is **No** as far as I know.

#### Is this a feature you're prepared to implement, with support from the npm CLI team?

I have never looked into npm code, but sure, I'd be happy to work on this.

**Comment by zkat** [comment]
My opinion is that this feature is largely superseded by `npx`, which takes care of multiple cases where people might have used global installs. It was, in fact, designed in such a way that people would never need to have global installs again, in almost all cases.

I suggest starting there. npx has been included with npm since `npm@5.2.0`, so you can expect it to be there as long as you have a fairly recent npm. You can read more about npx in [the medium post describing its use cases and features](https://medium.com/@maybekatz/introducing-npx-an-npm-package-runner-55f7d4bd282b).

**Comment by augnustin** [comment]
Thanks for feedback @zkat 

I did not know about the `npm bin` command. Pretty useful! The `npx` tool seems pretty neat too!

I'm also aware that many things around npm might work based on global packages (eg. `npm i -g npm`) so removing it would certainly not be an option. But what matters is how people *use* the feature.

So instead of removing the feature, I would rather discourage it. How?

By adding a warning, that would mention something like:

> Installing packages globally is discouraged unless you know what you are going. More info at https://docs.npmjs.com/troubleshooting/avoid-global-packages 
> Do you wish to continue? (Yes/No)

Then packages getting start wouldn't start with: 

> Getting started in one command: `npm install -g <PACKAGE_NAME>` 

anymore!

**Comment by zkat** [comment]
yeah I've thought about someday adding a warning for global packages, but I think it needs a bit longer to settle. In the meantime, we're just making sure local-only installs go as smoothly as possible and become more and more preferable.

Maybe actively discouraging globals is somewhere in the future. Who knows :)
